gobblin-kafka/src/main/java/gobblin/kafka/schemareg/CachingKafkaSchemaRegistry.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.kafka.schemareg;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+
+import com.google.common.base.Preconditions;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * An implementation that wraps a passed in schema registry and caches interactions with it
+ * {@inheritDoc}
+ * */
+@Slf4j
+public class CachingKafkaSchemaRegistry<K,S> implements KafkaSchemaRegistry<K,S> {
+
+  private static final int DEFAULT_MAX_SCHEMA_REFERENCES = 10;
+  private final KafkaSchemaRegistry<K,S> _kafkaSchemaRegistry;
+  private final HashMap<String, Map<S, K>> _namedSchemaCache;
+  private final HashMap<K, S> _idBasedCache;
+  private final int _maxSchemaReferences;
+
+
+  public CachingKafkaSchemaRegistry(KafkaSchemaRegistry kafkaSchemaRegistry)
+  {
+    this(kafkaSchemaRegistry, DEFAULT_MAX_SCHEMA_REFERENCES);
+  }
+
+  /**
+   * Create a caching schema registry.
+   * @param kafkaSchemaRegistry: a schema registry that needs caching
+   * @param maxSchemaReferences: the maximum number of unique references that can exist for a given schema.
+   */
+  public CachingKafkaSchemaRegistry(KafkaSchemaRegistry kafkaSchemaRegistry, int maxSchemaReferences)
+  {
+    Preconditions.checkArgument(kafkaSchemaRegistry!=null, "KafkaSchemaRegistry cannot be null");
+    if (kafkaSchemaRegistry.hasInternalCache())
+    {
+      log.warn(new StringBuilder().append("The passed in KafkaSchemaRegistry ")
```

Comment:
I am of the opinion that we shouldn't allow caching over a registry that has internal caching. It might be a reciepe for disaster.